src/fork_gc.c

```diff
@@ -0,0 +1,757 @@
+#include "fork_gc.h"
+#include "util/arr.h"
+#include "search_ctx.h"
+#include "inverted_index.h"
+#include "redis_index.h"
+#include "numeric_index.h"
+#include "tag_index.h"
+#include "tests/time_sample.h"
+#include <stdlib.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <wait.h>
+
+static void ForkGc_updateStats(RedisSearchCtx *sctx, ForkGCCtx *gc, size_t recordsRemoved,
+                               size_t bytesCollected) {
+  sctx->spec->stats.numRecords -= recordsRemoved;
+  sctx->spec->stats.invertedSize -= bytesCollected;
+  gc->stats.totalCollected += bytesCollected;
+}
+
+static void ForkGc_FDWriteLongLong(int fd, long long val) {
+  write(fd, &val, sizeof(long long));
+}
+
+static void ForkGc_FDWriteBuffer(int fd, char *buff, size_t len) {
+  ForkGc_FDWriteLongLong(fd, len);
+  if (len > 0) {
+    write(fd, buff, len);
+  }
+}
+
+static long long ForkGc_FDReadLongLong(int fd) {
+  long long ret;
+  ssize_t sizeRead = read(fd, &ret, sizeof(long long));
+  if (sizeRead != sizeof(long long)) {
+    return 0;
+  }
+  return ret;
+}
+
+static char *ForkGc_FDReadBuffer(int fd, size_t *len) {
+  *len = ForkGc_FDReadLongLong(fd);
+  if (*len == 0) {
+    return NULL;
+  }
+  char *buff = rm_malloc(*len * sizeof(char *));
+  read(fd, buff, *len);
+  return buff;
+}
+
+static bool ForkGc_InvertedIndexRepair(ForkGCCtx *gc, RedisSearchCtx *sctx, InvertedIndex *idx,
+                                       void (*RepairCallback)(const RSIndexResult *, void *),
+                                       void *arg) {
+  int *blocksFixed = array_new(int, 10);
+  int numDocsBefore[idx->size];
+  long long totalBytesCollected = 0;
+  long long totalDocsCollected = 0;
+  for (uint32_t i = 0; i < idx->size; ++i) {
+    IndexBlock *blk = idx->blocks + i;
+    if (blk->lastId - blk->firstId > UINT32_MAX) {
+      // Skip over blocks which have a wide variation. In the future we might
+      // want to split a block into two (or more) on high-delta boundaries.
+      // todo: is it ok??
+      continue;
+    }
+    IndexRepairParams params = {0};
+    params.RepairCallback = RepairCallback;
+    params.arg = arg;
+    numDocsBefore[i] = blk->numDocs;
+    int repaired = IndexBlock_Repair(blk, &sctx->spec->docs, idx->flags, &params);
+    // We couldn't repair the block - return 0
+    if (repaired == -1) {
+      return false;
+    }
+
+    if (repaired > 0) {
+      blocksFixed = array_append(blocksFixed, i);
+    }
+
+    totalBytesCollected += params.bytesCollected;
+    totalDocsCollected += repaired;
+  }
+
+  if (array_len(blocksFixed) == 0) {
+    // no blocks was repaired
+    ForkGc_FDWriteLongLong(gc->pipefd[1], 0);
+    array_free(blocksFixed);
+    return false;
+  }
+
+  // write number of repaired blocks
+  ForkGc_FDWriteLongLong(gc->pipefd[1], array_len(blocksFixed));
```

Comment:
It would be far more efficient to send all the information as one buffer rather than use micro-writes. I believe we have a Buffer or Array class to help with this.

Reply:
use something like gc->pipefd[GCREADER_FD] and [GC_WRITERFD], or similar